### PR TITLE
[lr] Add CosineAnnealing Learning Rate Scheduler

### PR DESCRIPTION
**Pull request by lhw414**
This PR adds the [CosineAnnealingLearningRateScheduler](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html) class, which adjusts the learning rate using a cosine annealing schedule. 

The implementation includes setting the max/min learning rates, decay steps, and proper error handling for missing properties.

### CODE REVIEW

**Comment by EunjuYang** [comment]
First of all, thank you for your contribution and PR.
We're going to review your PR. Before that, however, could you please check the CI issue? The NNTrainer's CI checks C++ formatting. 
For more information, please refer to this [page](https://github.com/nnstreamer/nntrainer/blob/main/docs/coding-convention.md).

**Review by lhs8928 — CHANGES_REQUESTED** [review]

File: nntrainer/optimizers/lr_scheduler_cosine.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  NNTR_THROW_IF(std::get<props::DecaySteps>(lr_props).empty(),
+                std::invalid_argument)
+    << "[CosineAnnealingLearningRateScheduler] Decay Steps is not set";
+  CosineAnnealingLearningRateScheduler::finalize();
+}
+
+void CosineAnnealingLearningRateScheduler::setProperty(
+  const std::vector<std::string> &values) {
+  auto left = loadProperties(values, lr_props);
+  CosineAnnealingLearningRateScheduler::setProperty(left);
```
[the remark this answers is not in the document]
Comment (baek2sm) [reply nntrainer/optimizers/lr_scheduler_cosine.cpp:44]: To add a little more explanation, property setting is completed while calling the loadProperties function, and the return value(left) contains values without matching property keys. So if there is a left value, you can make an exception.

You may modify it according to the following codes.
`void CosineAnnealingLearningRateScheduler::setProperty(
  const std::vector<std::string> &values) {
  auto left = loadProperties(values, lr_props);
  NNTR_THROW_IF(left.size(), std::invalid_argument)
    << "[CosineAnnealingLearningRateScheduler] There are unparsed properties";
}`

[the remark this answers is not in the document]
Comment (baek2sm) [reply nntrainer/optimizers/lr_scheduler_cosine.cpp:44]: To add a little more explanation, property setting is completed while calling the loadProperties function, and the return value(left) contains values without matching property keys. So if there is a left value, you can make an exception.

You may modify it according to the following codes.
`void CosineAnnealingLearningRateScheduler::setProperty(
  const std::vector<std::string> &values) {
  auto left = loadProperties(values, lr_props);
  NNTR_THROW_IF(left.size(), std::invalid_argument)
    << "[CosineAnnealingLearningRateScheduler] There are unparsed properties";
}`

**Review by baek2sm — CHANGES_REQUESTED** [review]

**Comment by baek2sm** [comment]
@DonghakPark For smooth code review process regarding this task, I have asked them to split the codes into smaller parts. Test codes will be uploaded later :)

File: nntrainer/optimizers/lr_scheduler_cosine.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lhw414) [reply nntrainer/optimizers/lr_scheduler_cosine.cpp]: There was a mistake during the implementation process while referencing multiple codes. It has been corrected to implement the functionality without calling methods from other classes.

File: nntrainer/optimizers/lr_scheduler_cosine.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+void CosineAnnealingLearningRateScheduler::exportTo(
+  Exporter &exporter, const ml::train::ExportMethods &method) const {
+  CosineAnnealingLearningRateScheduler::exportTo(exporter, method);
```
[the remark this answers is not in the document]
Comment (lhw414) [reply nntrainer/optimizers/lr_scheduler_cosine.cpp]: Same with setProperty().
I fixed it.

File: nntrainer/optimizers/lr_scheduler_cosine.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  exporter.saveResult(lr_props, method, this);
+}
+
+double CosineAnnealingLearningRateScheduler::getLearningRate(size_t iteration) {
+  auto const &max_lr = std::get<props::MaxLearningRate>(lr_props);
+  auto const &min_lr = std::get<props::MinLearningRate>(lr_props);
+  auto const &decay_steps = std::get<props::DecaySteps>(lr_props);
+
+  auto lr = CosineAnnealingLearningRateScheduler::getLearningRate(iteration);
```
[the remark this answers is not in the document]
Comment (lhw414) [reply nntrainer/optimizers/lr_scheduler_cosine.cpp]: Same with setProperty().
I fixed it.

File: nntrainer/optimizers/lr_scheduler_cosine.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  NNTR_THROW_IF(std::get<props::MaxLearningRate>(lr_props).empty(),
+                std::invalid_argument)
+    << "[CosineAnnealingLearningRateScheduler] Max Learning Rate is not set";
+  NNTR_THROW_IF(std::get<props::MinLearningRate>(lr_props).empty(),
+                std::invalid_argument)
+    << "[CosineAnnealingLearningRateScheduler] Min Learning Rate is not set";
+  NNTR_THROW_IF(std::get<props::DecaySteps>(lr_props).empty(),
+                std::invalid_argument)
+    << "[CosineAnnealingLearningRateScheduler] Decay Steps is not set";
+  CosineAnnealingLearningRateScheduler::finalize();
```
[the remark this answers is not in the document]
Comment (lhw414) [reply nntrainer/optimizers/lr_scheduler_cosine.cpp]: Same with setProperty().
I fixed it.

**Comment by lhw414** [comment]
I have a quick question. The [.clang-format file](https://github.com/nnstreamer/nntrainer/blob/main/.clang-format) does not have an option like InsertNewlineAtEOF: true, but TAOS's pre-build check verifies if there is a newline at the end of the file. I'm curious if this is because contributors generally use the "insert newline at EOF" option in their IDEs by default, or if there is another reason for the absence of such an option in clang-format.

**Comment by DonghakPark** [comment]
> @DonghakPark For smooth code review process regarding this task, I have asked them to split the codes into smaller parts. Test codes will be uploaded later :)

Thank you. By splitting commits, it will become much clearer to review them, and when looking at the history later on, we can clearly see the points where changes were made per feature.

**Review by baek2sm — APPROVED** [review]

**Review by EunjuYang — APPROVED** [review]
LGTM

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Review by lhs8928 — APPROVED** [review]
LGTM

**Comment by lhs8928** [comment]
How about we use full name "cosine annealing" for this learning rate instead of just "cosine"?
